spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -596,110 +617,208 @@ private void invokeListener(final ConsumerRecords<K, V> records) {
 			}
 		}
 
+		@SuppressWarnings("unchecked")
 		private void invokeBatchListener(final ConsumerRecords<K, V> records) {
 			List<ConsumerRecord<K, V>> recordList = new LinkedList<ConsumerRecord<K, V>>();
 			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
 			while (iterator.hasNext()) {
 				recordList.add(iterator.next());
 			}
 			if (recordList.size() > 0) {
-				try {
-					switch (this.listenerType) {
-						case ACKNOWLEDGING_CONSUMER_AWARE:
-							this.batchListener.onMessage(recordList,
-									this.isAnyManualAck
-											? new ConsumerBatchAcknowledgment(recordList)
-											: null, this.consumer);
-							break;
-						case ACKNOWLEDGING:
-							this.batchListener.onMessage(recordList,
-									this.isAnyManualAck
-											? new ConsumerBatchAcknowledgment(recordList)
-											: null);
-							break;
-						case CONSUMER_AWARE:
-							this.batchListener.onMessage(recordList, this.consumer);
-							break;
-						case SIMPLE:
-							this.batchListener.onMessage(recordList);
-							break;
-					}
-					if (!this.isAnyManualAck && !this.autoCommit) {
-						for (ConsumerRecord<K, V> record : getHighestOffsetRecords(recordList)) {
-							this.acks.put(record);
-						}
-					}
+				@SuppressWarnings("rawtypes")
+				Producer producer = null;
+				if (this.kafkaTxManager != null) {
+					producer = ProducerFactoryUtils
+							.getTransactionalResourceHolder(this.kafkaTxManager.getProducerFactory()).getProducer();
 				}
-				catch (Exception e) {
-					if (this.containerProperties.isAckOnError() && !this.autoCommit) {
-						for (ConsumerRecord<K, V> record : getHighestOffsetRecords(recordList)) {
-							this.acks.add(record);
-						}
+				if (this.transactionTemplate != null) {
+					invokeBatchListenerInTx(records, recordList, producer);
+				}
+				else {
+					doInvokeBatchListener(records, recordList, null);
+				}
+			}
+		}
+
+		private void invokeBatchListenerInTx(final ConsumerRecords<K, V> records,
+				List<ConsumerRecord<K, V>> recordList, @SuppressWarnings("rawtypes") Producer producer) {
+			try {
+				this.transactionTemplate.execute(s -> {
+					RuntimeException aborted = doInvokeBatchListener(records, recordList, producer);
+					if (aborted != null) {
+						throw aborted;
```

Comment:
Because, when we're not running in a transaction, we don't want to throw the exception; I suppose I could add `if (producer != null)` and throw instead of returning; let me know if you prefer that.